pkg/workflow/controllers/serial_node_reconciler.go

```diff
@@ -52,93 +65,224 @@ func (it *SerialNodeReconciler) Reconcile(request reconcile.Request) (reconcile.
-			return reconcile.Result{}, err
+			it.logger.Error(err, "failed to create child node",
+				"node", fmt.Sprintf("%s/%s", node.Namespace, node.Name),
+				"child node", childNode)
+			return err
 		}
+		childrenNames = append(childrenNames, childNode.Name)
+	}
+	it.logger.Info("serial node spawn new child node",
+		"node", fmt.Sprintf("%s/%s", node.Namespace, node.Name),
+		"child node", childrenNames)
 
-		for _, childNode := range childrenNodes {
-			err := it.kubeClient.Create(ctx, childNode)
-			if err != nil {
-				it.logger.Error(err, "failed to create child node", "node", request.NamespacedName, "child node", childNode)
-				return reconcile.Result{}, err
-			}
-		}
+	return nil
+}
 
-		updateError := retry.RetryOnConflict(retry.DefaultRetry, func() error {
-			nodeNeedUpdate := v1alpha1.WorkflowNode{}
-			err := it.kubeClient.Get(ctx, request.NamespacedName, &nodeNeedUpdate)
-			if err != nil {
-				return err
-			}
-			for _, item := range childrenNodes {
-				nodeNeedUpdate.Status.ActiveChildren = append(nodeNeedUpdate.Status.ActiveChildren, corev1.LocalObjectReference{Name: item.Name})
-			}
-			return it.kubeClient.Update(ctx, &nodeNeedUpdate)
-		})
+func (it *SerialNodeReconciler) fetchChildrenNodes(ctx context.Context, node v1alpha1.WorkflowNode) (activeChildrenNodes []v1alpha1.WorkflowNode, finishedChildrenNodes []v1alpha1.WorkflowNode, err error) {
+	childrenNodes := v1alpha1.WorkflowNodeList{}
+	controlledByThisNode, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			v1alpha1.LabelControlledBy: node.Name,
+		},
+	})
 
-		if updateError != nil {
-			return reconcile.Result{}, updateError
-		}
+	if err != nil {
+		it.logger.Error(err, "failed to build label selector with filtering children workflow node controlled by current node",
+			"current node", fmt.Sprintf("%s/%s", node.Namespace, node.Name))
+		return nil, nil, err
+	}
+
+	// TODO: sort with CreationTimestamp
+	err = it.kubeClient.List(ctx, &childrenNodes, &client.ListOptions{
+		LabelSelector: controlledByThisNode,
+	})
+
+	if err != nil {
+		it.logger.Error(err, "failed to list children workflow node controlled by current node",
+			"current node", fmt.Sprintf("%s/%s", node.Namespace, node.Name))
+		return nil, nil, err
+	}
+
+	sortedChildrenNodes := SortByCreationTimestamp(childrenNodes.Items)
+	sort.Sort(sortedChildrenNodes)
+
+	it.logger.V(4).Info("list children node", "current node",
+		"current node", fmt.Sprintf("%s/%s", node.Namespace, node.Name),
+		len(sortedChildrenNodes), "children", sortedChildrenNodes)
 
-		var childrenNames []string
-		for _, item := range childrenNodes {
-			childrenNames = append(childrenNames, item.Name)
+	var activeChildren []v1alpha1.WorkflowNode
+	var finishedChildren []v1alpha1.WorkflowNode
+
+	for _, item := range sortedChildrenNodes {
+		childNode := item
+		if WorkflowNodeFinished(childNode.Status) {
```

Comment:
If a chaos-node failed to apply the chaos, the `Finished` Condition could never be true, so that will pending the progress of workflow, which is expected. 😁

By the way, another condition `DeadlineExceed` will be not treated as finished after merging `nivara`.

We have not resolved the failed logic explicitly with all the workflow, such as sending some events or updating status.🙈
I think it should be introduced in another PR with a unified implementation.